blocks/model.py

```diff
@@ -0,0 +1,181 @@
+"""Defines models.
+
+A model is a thin layer of abstraction between the user-defined computation
+graph, bricks, parameters and main loop extensions. This module provides
+the basic :class:`AbstractModel` interface as well as its implementations
+(currently only :class:`Model`).
+
+"""
+import logging
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+from itertools import chain
+
+from six import add_metaclass
+
+from blocks.graph import ComputationGraph
+from blocks.select import Selector
+from blocks.filter import VariableFilter, get_brick
+from blocks.roles import PARAMETER
+
+logger = logging.getLogger()
+
+
+@add_metaclass(ABCMeta)
+class AbstractModel(object):
+    """A parameterized entity trained in the main loop.
+
+    A model is a parameterized entity the user trains a in a main loop.
+    The following are traits of every model:
+
+    * It has parameters and supports a way to access them. In addition
+    to returning handles to parameter objects it can return their values
+    as numpy arrays and set their values to given numpy arrays.
+
+    * It has an optimality criterion.
+
+    * It can be serialized and deserialized by mean of pickling.
+
+    * It might have bricks as its components.
+
+    This class provides an interface for models. For experiments use
+    a subclass, e.g. the :class:`Model`.
+
+    """
+    @abstractmethod
+    def get_params(self):
+        """Return the model parameters.
+
+        Returns
+        -------
+        params : OrderedDict
+            Dictionary of (name, parameter) pairs.
+
+        """
+        pass
+
+    def get_param_values(self):
+        """Return the values of model parameters.
+
+        The default implementation assumes that parameters are Theano
+        shared variables.
+
+        Returns
+        -------
+        param_values : OrderedDict
+            Dictionary of (parameter name, :class:`~numpy.ndarray`) pairs.
+
+        """
+        return OrderedDict((name, param.get_value())
+                           for name, param in self.get_params().items())
+
+    def set_param_values(self, param_values):
+        """Set the values of model parameters.
+
+        The default implementation assumes that parameters are Theano
+        shared variables.
+
+        Parameters
+        ----------
+        param_values : OrderedDict
+            Dictionary of (parameter name, :class:`~numpy.ndarray`) pairs.
+
+        """
+        params = self.get_params()
+
+        unknown = set(param_values) - set(params)
+        missing = set(params) - set(param_values)
+        if len(unknown):
+            logger.error("unknown parameter names:\n".format(unknown))
```

Comment:
On 18.02.2015 16:19, Bart van Merriënboer wrote:
>
> In blocks/model.py 
> <https://github.com/bartvm/blocks/pull/298#discussion_r24908298>:
>
> > +
> > +        The default implementation assumes that parameters are Theano
> > +        shared variables.
> > +
> > +        Parameters
> > +        ----------
> > +        param_values : OrderedDict
> > +            Dictionary of (parameter name, :class:`~numpy.ndarray`) pairs.
> > +
> > +        """
> > +        params = self.get_params()
> > +
> > +        unknown = set(param_values) - set(params)
> > +        missing = set(params) - set(param_values)
> > +        if len(unknown):
> > +            logger.error("unknown parameter names:\n".format(unknown))
>
> Haha, that's what I meant earlier when I said |basicConfig| is a 
> no-op. However, I think you're still right in saying that you 
> shouldn't just call |basicConfig| on import, because then if a user 
> does this:
>
> import  blocks.model
> logging.basicConfig(level='DEBUG')
>
> Nothing happens, which is confusing. This is because /libraries/ are 
> not expected to configure logging, but /applications/ are. The thing 
> is that we fall in either category depending on what part of the 
> library people are using (e.g. we should set a handler for the main 
> loop, but not when importing bricks). So basically, for Python 2 I 
> think we only want to call |basicConfig| at the point that we really 
> need a handler but there still isn't one.
>
Exactly.
>
> That allows users to configure the handlers when they want to.
>
> So, maybe we should just call |logging.basicConfig()| right before 
> |logger.error| calls?
>
That's something that should be done automatically, for which purpose we 
need something like ```logger = blocks_get_logger(__name__)``` in the 
beginning of each file. Out of the scope of this PR, I think.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/bartvm/blocks/pull/298/files#r24908298>.
>